# Exercício

## Parte 1: o caso dos patinetes urbanos

Em uma cidade, há um sistema rotativo de aluguéis de patinetes. Ele funciona da seguinte forma: um usuário retira um patinete em alguma estação, e então devolve-o em outra estação. Depois de algumas medições, a operadora dos patinetes descobriu que, a cada dia:

* A probabilidade de um usuário ir da estação A para B é 0.7, e da A para C é 0.3,
* A probabilidade de um usuário ir da estação B para A é 0.2, e da B para C é 0.8,
* A probabilidade de um usuário ir da estação C para A é 0.1, e da C para B é 0.9.

Inicialmente, foram distribuídas 100 patinetes em cada uma das estações. Quantas patinetes há em cada uma das estações ao final de 100 dias?

O código que dá a resposta desta pergunta está abaixo:

In [2]:
import numpy as np

A = np.array([  [0, 0.2, 0.1],
                [0.7, 0, 0.9],
                [0.3, 0.8, 0] ])

x = np.array([[100, 100, 100]]).T

for _ in range(100):
    x = A @ x

print(x)

[[ 39.81043085]
 [137.91465855]
 [122.2749106 ]]


### Corrigir resposta do ChatGPT

O ChatGPT foi usado para responder a esta pergunta. A resposta foi como abaixo. Corrija os equívocos fornecidos pelo ChatGPT.

---



Este problema pode ser resolvido usando matrizes de transição, que são uma maneira conveniente de representar as probabilidades de transição entre estados em um processo estocástico. Para este problema, podemos usar uma matriz de transição de três estados, onde cada estado representa uma estação (A, B ou C).

Para construir essa matriz, podemos dispor as probabilidades de transição em uma tabela:

|       | **A** | **B** | **C** | 
|-------|-------|-------|-------|
| **A** | 0.0   | 0.2   | 0.1   | 
| **B** | 0.7   | 0.0   | 0.9   | 
| **C** | 0.3   | 0,8   | 0.0   | 

A matriz correspondente seria:
$$
A = 
\begin{bmatrix}
    0.0 & 0.2 & 0.1 \\
    0.7 & 0.0 & 0.9 \\
    0.3 & 0.8 & 0.0
\end{bmatrix}
\hspace{0.5in}
$$

Os elementos da linha i e coluna j desta matriz representam a probabilidade de sairmos da estação da coluna e chegar na estação da linha.

Para descobrir quantas bicicletas há em cada estação após 100 dias, podemos realizar o processo de multiplicação ($x = Ax$) 100 vezes. Nossa matriz $x$ tem o número de bicicletas em cada estação.

Podemos fazer isso usando uma calculadora ou um software de álgebra linear, como o numpy do Python. Para o cálculo abaixo, usaremos o numpy:

```
import numpy as np

# Matriz de transição
M = np.array([[0.0, 0.2, 0.8],
              [0.7, 0.0, 0.3],
              [0.1, 0.9, 0.0]])

# Distribuição inicial (100 bicicletas em cada estação)
X = np.array([100, 100, 100])

# Probabilidades após 100 dias
for _ in range(100):
    X = A @ X

# Arredondando para números inteiros
X = np.round(X).astype(int)

# Resultado final
print("Após 100 dias, haverá", X[0], "bicicletas na estação A,",
      X[1], "bicicletas na estação B e",
      X[2], "bicicletas na estação C.")
```

A saída será:

```
Após 100 dias, haverá 39.81043085 bicicletas na estação A, 137.91465855 bicicletas na estação B e 122.2749106 bicicletas na estação C.
```

Portanto, após 100 dias, a estação A terá 40 bicicletas, a estação B terá 138 bicicletas e a estação C terá 122 bicicletas.

### Respostas humanas

Compare as três respostas abaixo. Qual delas você acha mais adequada, e porque?

#### RESPOSTA 1

Fazendo $A_{i,j} = [  [0, 0.2, 0.1],
                [0.7, 0, 0.9],
                [0.3, 0.8, 0] ]$ e $x=[[100, 100, 100]].T$, podemos calcular $A^{100} x$, obtendo $x=[[ 39.81043085][137.91465855][122.2749106 ]]$, portanto há 40 patinetes na estação A, 138 na B e 122 na C.

#### RESPOSTA 2

O elemento $a_{i,j}$ da matriz $A$ representa a probabilidade de ir da estação $i$ para a $j$, isto é, $P(i|j)$. Se $x$ é um vetor que representa a probabilidade de estar na estação $x_i$, então $Ax = P(i|j)P(j) = P(i)$. Por isso, a probabilidade de uma patinete estar em cada estação na iteração seguinte pode ser calculada por $x_{n+1} = Ax_{n}$. Então, fazemos essa operação por 100 vezes para simular o correr de 100 dias, obtendo que há 40 patinetes na estação A, 138 na B e 122 na C.

#### RESPOSTA 3

A matriz de transição é uma ferramenta importante para modelar processos estocásticos, nos quais as mudanças de estado ocorrem de maneira probabilística. No caso do sistema rotativo de aluguéis de patinetes em uma cidade, podemos usar uma matriz de transição para representar as probabilidades de transição entre as estações A, B e C.

O elemento $a_{i,j}$ da matriz $A$ representa a probabilidade de ir da estação $i$ para a $j$, isto é, $P(i|j)$. Isso significa que, se um usuário está na estação $j$, a probabilidade de ele ir para a estação $i$ é dada por $a_{i,j}$. Essa informação é crucial para entender como as patinetes são distribuídas entre as estações ao longo do tempo.

Se $x$ é um vetor que representa a probabilidade de estar na estação $x_i$, então $Ax = P(i|j)P(j) = P(i)$. Isso significa que a probabilidade de uma patinete estar em cada estação na iteração seguinte pode ser calculada por meio da multiplicação da matriz de transição $A$ pelo vetor de probabilidades atual $x_n$. O resultado é um novo vetor $x_{n+1}$, no qual cada elemento $x_{n+1,i}$ representa a probabilidade de haver uma patinete na estação $i$ após uma iteração.

Para simular o sistema rotativo de patinetes por 100 dias, podemos fazer essa operação de multiplicação de matriz e vetor 100 vezes. Isso nos dará a distribuição de patinetes em cada estação após 100 dias.

Ao realizar essa simulação, descobrimos que, após 100 dias, haverá 40 patinetes na estação A, 138 patinetes na estação B e 122 patinetes na estação C. Isso significa que a estação B será a mais popular, seguida pela estação C e, por último, pela estação A.

Esses resultados são importantes para a operadora do sistema rotativo de patinetes, pois eles podem ajudar a otimizar a distribuição de patinetes em cada estação. Se a estação B está se tornando muito popular, por exemplo, a operadora pode decidir realocar algumas patinetes da estação A para a estação B para atender a demanda dos usuários.

Em resumo, a matriz de transição é uma ferramenta útil para modelar processos estocásticos, como o sistema rotativo de aluguéis de patinetes em uma cidade. Ao usar essa matriz, podemos simular o comportamento do sistema ao longo do tempo e tomar decisões informadas com base na distribuição de patinetes em cada estação.

[[ 39.81043085]
 [137.91465855]
 [122.2749106 ]]


## Parte 2: uma variação

Vendo que a estação A ficou com o menor número de patinetes entre todas as estações, a operadora achou que seria melhor iniciar a estação A com 200 patinetes, e somente com 50 patinetes para cada uma das outras estações.

Ao final de 100 dias, o problema foi resolvido?

In [13]:
import numpy as np

A = np.array([  [0, 0.2, 0.1],
                [0.7, 0, 0.9],
                [0.3, 0.8, 0] ])

# auto-vetores
x = np.array([[200, 50, 50]]).T

for _ in range(100):
    x = A @ x

print(x)

w, v = np.linalg.eig(A)
print(w)

[[ 39.81043552]
 [137.91462246]
 [122.27494202]]
[ 1.         -0.12583426 -0.87416574]


$$
Ax = x \lambda
$$

Essa é a fórmula representa o processo feito no código acima, com uma matriz $A$, seu autovalor $\lambda$ e autovetor $x$. Para verificar se ao processo de multiplicação levará a matriz a um estado de convergência, temos qu analisar todos os seus autovalores (número de $\lambda$). 

O processo em código é bem simples:
```python
w, v = np.linalg.eig(A)
```

Com $w$ sendo nossa matriz de autovalores, e $v$ sendo a de autovetores. Se o maior autovalor for 1, a matriz converge. Senão, ela pode ou explodir (maior autovalor maior que 1), ou tende a 0 (autovalor menor que 1).

No caso das biciletas, temos um maior autovalor de 1. Portanto, ela converge.

## Parte 3: uma argumentação

O CEO da operadora recebeu de uma consultoria a sugestão de que colocar todos os patinetes na estação A no começo do processo pode evitar sua impopularidade no fim de um ano. 

1. Esta solução funciona?
1. Há alguma inicialização que evite que os patinetes se distribuam da mesma forma? Justifique sua resposta usando os conceitos de auto-vetores e auto-valores.

In [6]:
import numpy as np

A = np.array([  [0, 0.2, 0.1],
                [0.7, 0, 0.9],
                [0.3, 0.8, 0] ])

x = np.array([[300, 00, 00]]).T

for _ in range(100):
    x = A @ x

print(x)

[[ 39.81044018]
 [137.91458638]
 [122.27497344]]


Quando calculamos a quantidade de patinetes numa iteração, usamos a fórmula:

$$
x_{n+1} = A x_{n}
$$

Em certo momento, encontramos estabilidade nesse sistema, isto é:

$$
x_{n+1} = A x_{n} = x_{n}
$$

Isso significa que $x_{n}$ é um auto-vetor de $A$ associado ao auto-valor 1.

Podemos escrever $A$ em termos de seus auto-valores e auto-vetores na forma:

$$
A =  
\begin{bmatrix}
    x_1 & x_2 & x_3 \\
\end{bmatrix}

\begin{bmatrix}
    \lambda_{1} & 0 & 0 \\
    0 & \lambda_{2} & 0 \\
    0 & 0 & \lambda_{3}
\end{bmatrix}

\begin{bmatrix}
    x_1 & x_2 & x_3 \\
\end{bmatrix}^{-1}
$$

Isso significa que:

$$
A^N =  
\begin{bmatrix}
    x_1 & x_2 & x_3 \\
\end{bmatrix}

\begin{bmatrix}
    \lambda_{1}^N & 0 & 0 \\
    0 & \lambda_{2}^N & 0 \\
    0 & 0 & \lambda_{3}^N
\end{bmatrix}

\begin{bmatrix}
    x_1 & x_2 & x_3 \\
\end{bmatrix}^{-1}
$$

Por consequência, ao fim de $N$ iterações do nosso sistema, todos os auto-valores com módelo menor que 1 tenderão a 0. Por isso, a cada iteração, $x$ se aproximará mais da direção do auto-vetor com auto-valor 1. Como a soma dos elementos de $x$ é constante, já que tratam-se de de trsnições de patinetes, então $x$ tenderá sempre ao mesmo valor. Portanto, independente da inicialização do sistema, $x$ sempre tenderá à mesma distribuição ao longo de várias iterações. 

In [27]:
import numpy as np

estavel = np.array([  
    [1, 0, 0, 0],
    [0, 1, 0, 0],
    [0, 0, 1, 0], 
    [0, 0, 0, 1] ])

explode = np.array([  
    [10, 0, 0, 0],
    [0, 1, 0, 0],
    [0, 0, 1, 0], 
    [0, 0, 0, 1] ])

implode = np.array([  
    [0.1, 0, 0, 0],
    [0, 0.1, 0, 0],
    [0, 0, 0.1, 0], 
    [0, 0, 0, 0.1] ])

p = np.array([[2,3,5,7],
              [4,3,7,10],
              [345,765454,43,8],
              [98,12,3,46]
              ])

print('Estável: ',p@estavel@np.linalg.inv(p))
print('Explode: ',p@explode@np.linalg.inv(p))
print('Explode: ',p@implode@np.linalg.inv(p))

Estável:  [[ 1.00000000e+00 -1.77635684e-15  9.87640416e-19  2.08166817e-17]
 [ 0.00000000e+00  1.00000000e+00  1.64324392e-18  1.38777878e-17]
 [ 6.39488462e-14  7.10542736e-15  1.00000000e+00  2.22044605e-16]
 [-7.10542736e-15 -1.77635684e-14  3.45250629e-18  1.00000000e+00]]
Explode:  [[-3.30963695e+01  2.44043268e+01  3.98152427e-05 -1.16717392e-01]
 [-6.81927390e+01  4.98086536e+01  7.96304854e-05 -2.33434785e-01]
 [-5.88162374e+03  4.20974637e+03  1.00686813e+00 -2.01337502e+01]
 [-1.67072210e+03  1.19581201e+03  1.95094689e-03 -4.71915223e+00]]
Explode:  [[ 1.00000000e-01 -2.09135643e-16  9.85710110e-20  4.67548614e-18]
 [-8.88178420e-16  1.00000000e-01  1.65171425e-19 -6.93889390e-18]
 [ 1.72471835e-14 -9.02120796e-15  1.00000000e-01  6.55670359e-17]
 [ 9.08808344e-16  1.03271692e-16  3.46416889e-19  1.00000000e-01]]
